src/imitation/data/huggingface_datasets_conversion.py

```diff
@@ -0,0 +1,135 @@
+"""Helpers to convert between Trajectories and HuggingFace's datasets library."""
+import functools
+from typing import Any, Dict, Iterable, Sequence, cast
+
+import datasets
+import jsonpickle
+import numpy as np
+
+from imitation.data import types
+
+
+class TrajectoryDatasetSequence(Sequence[types.Trajectory]):
+    """A wrapper to present a HF dataset as a sequence of trajectories.
+
+    Converts the dataset to a sequence of trajectories on the fly.
+    """
+
+    def __init__(self, dataset: datasets.Dataset):
+        """Construct a TrajectoryDatasetSequence."""
+        # TODO: this is just a temporary workaround for
+        #  https://github.com/huggingface/datasets/issues/5517
+        #  switch to .with_format("numpy") once it's fixed
+        def numpy_transform(batch):
+            return {key: np.asarray(val) for key, val in batch.items()}
+
+        self._dataset = dataset.with_transform(numpy_transform)
+        self._trajectory_class = (
+            types.TrajectoryWithRew if "rews" in dataset.features else types.Trajectory
+        )
+
+    def __len__(self) -> int:
+        return len(self._dataset)
+
+    def __getitem__(self, idx):
+
+        if isinstance(idx, slice):
+            dataslice = self._dataset[idx]
+
+            # Extract the trajectory kwargs from the dataset slice
+            trajectory_kwargs = [
+                {key: dataslice[key][i] for key in dataslice}
+                for i in range(len(dataslice["obs"]))
+            ]
+
+            # Ensure that the infos are decoded lazily using jsonpickle
+            for kwargs in trajectory_kwargs:
+                kwargs["infos"] = _LazyDecodedList(kwargs["infos"])
+
+            return [self._trajectory_class(**kwargs) for kwargs in trajectory_kwargs]
+        else:
+            # Extract the trajectory kwargs from the dataset
+            kwargs = self._dataset[idx]
+
+            # Ensure that the infos are decoded lazily using jsonpickle
+            kwargs["infos"] = _LazyDecodedList(kwargs["infos"])
+
+            return self._trajectory_class(**kwargs)
+
+
+class _LazyDecodedList(Sequence[Any]):
+    """A wrapper to lazily decode a list of jsonpickled strings.
+
+    Decoded results are cached to avoid decoding the same string multiple times.
+
+    This is used to decode the infos of a trajectory only when they are accessed.
+    """
+
+    def __init__(self, encoded_list: Sequence[str]):
+        self._encoded_list = encoded_list
+
+    def __len__(self):
+        return len(self._encoded_list)
+
+    # arbitrary cache size just to put a limit on memory usage
+    @functools.lru_cache(maxsize=100000)
+    def __getitem__(self, idx):
+        if isinstance(idx, slice):
+            return [jsonpickle.decode(info) for info in self._encoded_list[idx]]
+        else:
+            return jsonpickle.decode(self._encoded_list[idx])
+
+
+def trajectories_to_dataset(
```

Comment:
I guess this is beyond the scope of reviewing the PR, but why is this better than pickling a dataclass? A compelling case for this would be very relevant for other current projects.

Reply:
Valid question I think! For me there are the following points:

- It is generally not a good idea to serialize classes for persistence. Whenever you change the class implementation, you might (silently) loose the ability to load older files. I had this happen when loading SB3 models trained with old gym versions (with gym changing implementations with no regard to backwards compatibility)
- The pickle format changes from time to time and sometimes I have to mess with pickle versions.
- The HuggingFace Datasets are backed by Apache Arrow. This is mostly relevant in case our datasets grow. It allows to stream datasets (from disk or from the internet) and does memory mapping for us. With the memory mapping we can load huge files that would never fit in RAM and let the OS do transparent caching for us.
- The HuggingFace datasets do not support pickled objects by default (probably due to the above reasons and to make users adhere to best practices). We could make it happen with a additional loading script next to the pickled classes but this would increase complexity and I really don't like storing code together with our datasets.

You can read more about the considerations in #651